Initial attempt at documenting MMM01 (#310)
gbdev/pandocs: src/MMM01.md

# MMM01

> The MMM01 is a mapper specific to multi-game compilation cartridges. It emulates an MBC1 for the contained games, and supports containing a mix of games from 32 kiB ROM with no RAM up to the same maximum memory _per game_ as the MBC1:

Author: Lower case "k" is correct for "kilo"

> The MMM01 is a mapper specific to multi-game compilation cartridges. It emulates an MBC1 for the contained games, and supports containing a mix of games from 32 kiB ROM with no RAM up to the same maximum memory _per game_ as the MBC1:

Reviewer: Except that [this is not kilo, but "kibi"](https://en.wikipedia.org/wiki/Binary_prefix#kibi), for which both "k" and "K" are accepted:

> The first letter of each such prefix is therefore identical to the corresponding SI prefixes, except for "K", which is used interchangeably with "k", whereas in SI, only the lower-case k represents 1000.

[The style guide recommends uppercase K](https://github.com/gbdev/pandocs/wiki/Document-Style#4-units-and-prefixes), too.

> The MMM01 is a mapper specific to multi-game compilation cartridges. It emulates an MBC1 for the contained games, and supports containing a mix of games from 32 kiB ROM with no RAM up to the same maximum memory _per game_ as the MBC1:

Author: Ok that's fair. I'll correct it.

> The MMM01 is a mapper specific to multi-game compilation cartridges. It emulates an MBC1 for the contained games, and supports containing a mix of games from 32 kiB ROM with no RAM up to the same maximum memory _per game_ as the MBC1:

Reviewer: I noticed I missed some instances, so please correct these as well. (A case-sensitive `\bkiB\b` regex should take care of it pretty quickly, actually.)

* max 512 kiByte ROM with banked 8, 16, or 32 kiByte RAM (default configuration)
* max 2 MiByte ROM with unbanked 8 kiByte RAM ("multiplex" mode[^multiplex])

[^multiplex]:
However, no cartridges were released using the MMM01 "multiplex" mode.

Regardless of the size or number of the games in the compilation, the maximum total cartridge size supported by the MMM01 is the same - up to 8 MiB ROM and 128 kiB RAM.

> Additional bits in the control registers extend the MBC1 ROM and RAM banking numbers to allow for game selection, and the MBC1 ROM/RAM banking registers can be masked so some of _those_ bits can also be used for game selection, allowing for up to 255x 32 kiB games, plus a 32kiB menu, in an 8 MiB ROM. RAM is more limited at only up to 16x 8kB RAM banks in total. However, despite these generous capabilities, no MMM01 cartridge was released with more than 4 games, and only _one_ contains any RAM at all.

Author: That's not strictly accurate - the extra bits can _only_ be used for game selection. It's the lower MBC1-compatible bits that are maskable to be either for game selection or in-game banking use.

> Additional bits in the control registers extend the MBC1 ROM and RAM banking numbers to allow for game selection, and the MBC1 ROM/RAM banking registers can be masked so some of _those_ bits can also be used for game selection, allowing for up to 255x 32 kiB games, plus a 32kiB menu, in an 8 MiB ROM. RAM is more limited at only up to 16x 8kB RAM banks in total. However, despite these generous capabilities, no MMM01 cartridge was released with more than 4 games, and only _one_ contains any RAM at all.

Author: Also, the extra bits aren't all in the "correct" registers, so the distinction has to be made between bank number and register

> Additional bits in the control registers extend the MBC1 ROM and RAM banking numbers to allow for game selection, and the MBC1 ROM/RAM banking registers can be masked so some of _those_ bits can also be used for game selection, allowing for up to 255x 32 kiB games, plus a 32kiB menu, in an 8 MiB ROM. RAM is more limited at only up to 16x 8kB RAM banks in total. However, despite these generous capabilities, no MMM01 cartridge was released with more than 4 games, and only _one_ contains any RAM at all.

Reviewer: Depends on what you consider the "primary" register, I think: are the different bit groups individual registers, which are combined (and transformed) to produce a ROM bank number, or are they just separated views of a single, larger ROM bank register?

I prefer the latter for uniformity, but external opinions are welcome.

> Additional bits in the control registers extend the MBC1 ROM and RAM banking numbers to allow for game selection, and the MBC1 ROM/RAM banking registers can be masked so some of _those_ bits can also be used for game selection, allowing for up to 255x 32 kiB games, plus a 32kiB menu, in an 8 MiB ROM. RAM is more limited at only up to 16x 8kB RAM banks in total. However, despite these generous capabilities, no MMM01 cartridge was released with more than 4 games, and only _one_ contains any RAM at all.

Author: In the other MBC docs, there's a "registers" header which refers to them by address ranges, and the same is true here - so there's a precedence that "register" refers to the written values. Particularly noteworthy is the MBC1 doc, where the second register can also be used for extending the ROM banking number. This gives a precedence that the bank number is not the bank register.

The ROM and RAM "game select" bank numbers do not have to be set to the same number - this allows an MMM01 cartridge to not waste RAM space on games that do not have RAM[^no_ram], or mix and match games that have different sized ROM or RAM by packing them in tightly in the overall ROM/RAM of the cartridge.

[^no_ram]:
Technically, the MMM01 can't completely block access to RAM for a game, so if the cartridge contains RAM it's recommended to assign any no-RAM games to the same single RAM bank to prevent no-RAM games from accessing or corrupting other games' SRAM saves.

> MMM01 starts up in a unique mode ("unmapped") which always maps the _last 32 kiB_ of the ROM to the 0000-7FFF address region regardless of the values in the bank/mode registers. The correct ROM header (with Nintendo logo) therefore needs to be located at offset `(size - 32 kiB) + 100h` in the ROM rather than the usual `0000 + 100h` (which contains the header of the first game in the collection instead). MMM01 cartridges have a simple menu program in this last 32 kiB, which manipulates the additional MMM01 control bits, allowing game selection and setting the game size, before entering "mapped" mode and booting the selected game (see "Mapping Enable" below).

Author: Pandocs were originally as much for homebrew development as emulator development - that last line is relevant for homebrew.

As the last 32 kiB of the ROM are reserved for the cartridge menu, it's best to pack games into the ROM from largest to smallest to avoid having a game overlap the menu. e.g. the Taito Variety Pack contains three 128 kiB games and one 64 kiB game, leaving 32 kiB unused and 32 kiB for the menu in a 512 kiB ROM chip.

## Memory

> ### 0000-3FFF - ROM Bank X0 (Read Only)

Author: The style guide says no $ on addresses

> ### 0000-3FFF - ROM Bank X0 (Read Only)

Author: "ROM Bank X0" (with no $) is copied off the MBC1 docs - I've checked and none of the other MBC docs use $ on the bank number in these headers either

> ### 0000-3FFF - ROM Bank X0 (Read Only)

Reviewer: OK for the addresses, the context is clear enough from the "3FFF"; but I don't think it is for the "X0", so I insist on the $ there. The other docs don't have it because nobody has bothered to fix them yet—Pan Docs is full of such minor errors that have yet to be corrected.

### A000-BFFF - RAM Bank 00-03, if any (Read/Write)

This area is used to address external SRAM in the cartridge (if any). The RAM is only accessible if RAM is enabled, otherwise reads return open bus values (often $FF, but not guaranteed) and writes are ignored.

External RAM is often battery-backed, allowing for the storage of game data while the Game Boy is turned off, or if the cartridge is removed from the Game Boy.

It is currently unknown whether RAM access is possible while in unmapped mode.

## Registers

The MMM01 registers are extensions of the MBC1 registers. In "mapped" mode, the extra bits are unwriteable and the MMM01 emulates an MBC1. In "unmapped" mode, the extra bits are writeable, allowing the menu program to select which game to play and configure its size and RAM access.

All the MMM01 registers are 7-bits in size and set to 00h on power-up. For the ROM Bank Number register, this behaves as if it was set to 01h.

### 0000-1FFF - RAM Enable (Write Only)

```

> Bits: X 6 5 4 3 2 1 0

Author: Can you show me an example of what you mean?

> Bits: X 6 5 4 3 2 1 0

Reviewer: [This](https://gbdev.io/pandocs/STAT.html#ff41---stat-lcd-status-rw).

> Bits: X 6 5 4 3 2 1 0

Author: That's... really ugly.

> Bits: X 6 5 4 3 2 1 0

Reviewer: [We know.](https://github.com/gbdev/pandocs/issues/89)

> Bits: X 6 5 4 3 2 1 0

Author: Can we not keep the diagrams? The table loses the "visual" aspect of it.

I copied the concept from the tauwasser's MMM01 docs and found it very readable there.

> Bits: X 6 5 4 3 2 1 0

Reviewer: If you think the format should be changed (and yes, monospaced text must definitely be removed for this purpose), please open an issue so that we can decide on what replacement is best.

In the meantime, consistency is more important, if only so as to make conversion later on less painful.

> Bits: X 6 5 4 3 2 1 0

Author: I made one: #318 (I can't seem to label it as "consistency/style")

> Bits: X 6 5 4 3 2 1 0

Reviewer: That's because only maintainers can modify labels, of course ;)

> Bits: X 6 5 4 3 2 1 0

Reviewer: While we wait for #318 to be resolved (which will entail making some custom markup, so it'll take time), can those be changed to the same template as the rest of the document? They'll be converted in due time, and it's not worth letting this PR rot just for this.

> Bits: X 6 5 4 3 2 1 0

Author: I am currently away

        | \_/ \_____/
        |  |      \----- Bits 0-3: RAM Enable
        |  \------------ Bits 4-5: RAM Bank Mask
        \--------------- Bit 6: Mapping Enable
```

#### Bits 0-3: RAM Enable

As per MBC1, writing 0Ah here enables the RAM, and any other value disables the RAM. RAM is automatically disabled when the gameboy is powered off.

#### Bits 4-5: RAM Bank Mask

Can only be written in unmapped mode.

> If either or both of these bits are set high, the corresponding bit(s) in the MBC1 RAM banking register are locked in as "game select" bits instead of "banking" bits. Any attempt to write to those bits while its matching bit in this register is 1 is prevented.

Author: It applies all the time. I'll add that.

#### Bit 6: Mapping Enable

Can only be written in unmapped mode.

If 1, the MMM01 enters "mapped" mode. This immediately begins MBC1 emulation, mapping in the selected banks of the game ROM (typically 00h and 01h within the game's subsection of the cartridge ROM) and prevents write access to any of the MMM01 extended control bits.

> It is unknown if setting bit 6 to enter mapping mode will prevent simultaneous writes to bits 4-5 (RAM Bank Mask). The only released MMM01 cartridge containing RAM performs two separate writes to set the bank mask before enabling mapping.

Author: yes

### 2000-3FFF - ROM Bank Number (Write Only)

```
Bits: X 6 5 4 3 2 1 0
        \_/ \_______/
         |       \------ Bits 0-4: ROM Bank Low
         \-------------- Bits 5-6: ROM Bank Mid
```

#### Bits 0-4: ROM Bank Low

> This is equivalent to the MBC1 ROM Bank register, and (when in mapped mode, in addition to the ROM Bank Mid/High game select registers) selects the ROM bank visible in the 4000-7FFF region. It can be masked to reduce its size, reserving some of the bits for game select (see "ROM Bank Mask"). Bits masked for game select also affect the bank visible in the 0000-3FFF region.

Author: I've moved it

> If the _unmasked_ bits are all 0, it behaves as if the lowest bit is set to 1 (as per MBC1 behaviour, mapping bank 00 maps bank 01). However, the actual value of the register doesn't change.

Author: I've attempted to explain it in the mask section, but here's an example:

$10 written to ROM Bank Low
$10 written to ROM Bank Mask
// entering mapped mode at this point would map actual bank $11 to 4000-7FFF because the unmasked bits are all 0s

So unlike MBC1, the remapping isn't only triggered by a write of $00 to the ROM Bank register. I believe MBC1 is the same (only using a 1, but storing it as 0) but the difference is not observable on an actual MBC1.

> If the _unmasked_ bits are all 0, it behaves as if the lowest bit is set to 1 (as per MBC1 behaviour, mapping bank 00 maps bank 01). However, the actual value of the register doesn't change.

Reviewer: Ok, yeah, it's observable, so definitely backed up. Could this rationale be noted? For example:
```suggestion
If all _unmasked_ bits equal 0, the lowest bit, which is always writable, behaves as if set to 1 (as per MBC1 behaviour, mapping bank $00 maps bank $01).
However, the actual value of the register doesn't change, as evidenced if changing the [masking register](...) to un-trigger this behavior.
```

> Note: It's believed the lowest bit of the bank number (post 00->01 remapping) still affects the bank mapped to the 4000-7FFF region in "unmapped" mode - causing it to swap between the same second-last bank as 0000-3FFF (if the unmasked bits of ROM Bank Low are set to a non-zero value with the lowest bit 0) or the last bank of the ROM (if the unmasked bits of ROM Bank Low are all zero (which is treated as 01) or the lowest bit is in fact 1).

Author: The new version has this:

```
Note: It's suspected the lowest bit of ROM Bank Low (post $00->$01 remapping) still affects the bank mapped to the 4000-7FFF region in "unmapped" mode the same as it does in mapped mode.
Most of the time this would still be a 1, but during game selection it could momentarily go to 0 in between setting the game select bits in [ROM Bank Low](<#Bits 0-4: ROM Bank Low>) and masking them as such in [ROM Bank Mask](<#Bits 1-5: ROM Bank Mask>).
```

Is that better? I've removed a bit of technical detail as the behaviour seems to be unverified.

> Note: It's believed the lowest bit of the bank number (post 00->01 remapping) still affects the bank mapped to the 4000-7FFF region in "unmapped" mode - causing it to swap between the same second-last bank as 0000-3FFF (if the unmasked bits of ROM Bank Low are set to a non-zero value with the lowest bit 0) or the last bank of the ROM (if the unmasked bits of ROM Bank Low are all zero (which is treated as 01) or the lowest bit is in fact 1).

Reviewer: Okay, but now I don't know what it's trying to imply. That the penultimate 16 KiB bank may be mapped to ROMX in "unmapped" mode under some circumstances?

> Note: It's believed the lowest bit of the bank number (post 00->01 remapping) still affects the bank mapped to the 4000-7FFF region in "unmapped" mode - causing it to swap between the same second-last bank as 0000-3FFF (if the unmasked bits of ROM Bank Low are set to a non-zero value with the lowest bit 0) or the last bank of the ROM (if the unmasked bits of ROM Bank Low are all zero (which is treated as 01) or the lowest bit is in fact 1).

Author: Yeah. Basically the suspicion is that the low bit isn't forces to 1 by any "unmapped mode" logic, and just keeps its normal value from the normal banking logic.

This can mean it's 0 in some situations

> Note: It's believed the lowest bit of the bank number (post 00->01 remapping) still affects the bank mapped to the 4000-7FFF region in "unmapped" mode - causing it to swap between the same second-last bank as 0000-3FFF (if the unmasked bits of ROM Bank Low are set to a non-zero value with the lowest bit 0) or the last bank of the ROM (if the unmasked bits of ROM Bank Low are all zero (which is treated as 01) or the lowest bit is in fact 1).

Author: Yeah. Basically the suspicion is that the low bit doesn't have any special value in "unmapped mode", and just keeps its normal value from the normal banking logic.

This can mean it's 0 in some situations.

> Note: It's believed the lowest bit of the bank number (post 00->01 remapping) still affects the bank mapped to the 4000-7FFF region in "unmapped" mode - causing it to swap between the same second-last bank as 0000-3FFF (if the unmasked bits of ROM Bank Low are set to a non-zero value with the lowest bit 0) or the last bank of the ROM (if the unmasked bits of ROM Bank Low are all zero (which is treated as 01) or the lowest bit is in fact 1).

Author: Which means it ends up mapping bank 1FE to _both_ ROM areas

> Note: It's believed the lowest bit of the bank number (post 00->01 remapping) still affects the bank mapped to the 4000-7FFF region in "unmapped" mode - causing it to swap between the same second-last bank as 0000-3FFF (if the unmasked bits of ROM Bank Low are set to a non-zero value with the lowest bit 0) or the last bank of the ROM (if the unmasked bits of ROM Bank Low are all zero (which is treated as 01) or the lowest bit is in fact 1).

Reviewer: So what you're saying is that all that "unmapped" mode does, is treat all bank bits *but bit 0* as high?

> Note: It's believed the lowest bit of the bank number (post 00->01 remapping) still affects the bank mapped to the 4000-7FFF region in "unmapped" mode - causing it to swap between the same second-last bank as 0000-3FFF (if the unmasked bits of ROM Bank Low are set to a non-zero value with the lowest bit 0) or the last bank of the ROM (if the unmasked bits of ROM Bank Low are all zero (which is treated as 01) or the lowest bit is in fact 1).

Author: It's believed so yes. That and enables write access to the extra MMM01 control bits.

But my only source for that is tanhauser's document, which is confusing at best in places.

> Note: It's believed the lowest bit of the bank number (post 00->01 remapping) still affects the bank mapped to the 4000-7FFF region in "unmapped" mode - causing it to swap between the same second-last bank as 0000-3FFF (if the unmasked bits of ROM Bank Low are set to a non-zero value with the lowest bit 0) or the last bank of the ROM (if the unmasked bits of ROM Bank Low are all zero (which is treated as 01) or the lowest bit is in fact 1).

Reviewer: Then I would rather word it that way, and explain this caveat at the section about "unmapped" mode, instead of here.

#### Bits 5-6: ROM Bank Mid

_Can only be written in unmapped mode._

This register represents an additional two bits of ROM bank number, for game selection. Affects both the 0000-3FFF and 4000-7FFF region. Can only be used for game selection, as it's not writeable once entering a game (mapped mode).

In multiplexed mode, functionality is swapped with RAM Bank Low (See "Multiplex Enable") allowing for larger game ROM.

### 4000-5FFF - RAM Bank Number (Write Only)

Bits: X 6 5 4 3 2 1 0
        | \_/ \_/ \_/
        |  |   |   \---- Bits 0-1: RAM Bank Low
        |  |   \-------- Bits 2-3: RAM Bank High
        |  \------------ Bits 4-5: ROM Bank High
        \--------------- Bit 6: MBC1 Mode Write Disable

#### Bits 0-1: RAM Bank Low

> This is equivalent to the MBC1 RAM Bank register. It can be masked to reduce its size, reserving some of the bits for game select (see "RAM Bank Mask").

Author: I've added an xref for RAM Bank Mask, how can I xref the MBC1 RAM Bank register from another file?

> This is equivalent to the MBC1 RAM Bank register. It can be masked to reduce its size, reserving some of the bits for game select (see "RAM Bank Mask").

Reviewer: Just use `[MBC RAM Bank register](<#Name of the section>)`, [as explained in the README](https://github.com/gbdev/pandocs#internal-links).

> This is equivalent to the MBC1 RAM Bank register. It can be masked to reduce its size, reserving some of the bits for game select (see "RAM Bank Mask").

Author: Does that work between files?

> This is equivalent to the MBC1 RAM Bank register. It can be masked to reduce its size, reserving some of the bits for game select (see "RAM Bank Mask").

Reviewer: Yep!

> This is equivalent to the MBC1 RAM Bank register. It can be masked to reduce its size, reserving some of the bits for game select (see "RAM Bank Mask").

Author: Yes apparently.

We have the same subsection headings for bank registers in multiple MBC documents - how do we disambiguate which one we're linking to?

> This is equivalent to the MBC1 RAM Bank register. It can be masked to reduce its size, reserving some of the bits for game select (see "RAM Bank Mask").

Reviewer: ...that's annoying. It'll probably pick up the first section thusly named; we'll have to either change the section names in those pages to become unique, or manually link to the page, anchor included. Maybe I should introduce alternative syntax for this.

> This is equivalent to the MBC1 RAM Bank register. It can be masked to reduce its size, reserving some of the bits for game select (see "RAM Bank Mask").

Author: Does _this_ work? https://www.markdownguide.org/extended-syntax/#heading-ids

> This is equivalent to the MBC1 RAM Bank register. It can be masked to reduce its size, reserving some of the bits for game select (see "RAM Bank Mask").

Reviewer: No, it doesn't.

This register represents an additional two bits of RAM bank number, for game selection. Can only be used for game selection, as it's not writeable once entering a game (mapped mode).

#### Bits 4-5: ROM Bank High

_Can only be written in unmapped mode._

This register represents _another_ two bits of ROM bank number, for game selection. Affects both the 0000-3FFF and 4000-7FFF region. Can only be used for game selection, as it's not writeable once entering a game (mapped mode).

#### Bit 6: MBC1 Mode Write Disable

_Can only be written in unmapped mode._

When set to 1, prevents changes to the "MBC1 Mode Select". This might be for compatibility with games designed for non-MBC1 mappers that don't expect the MBC1 mode select register to exist.

### 6000-7FFF - Banking Mode Select (Write Only)

```
Bits: X 6 5 4 3 2 X 0
        | \_______/ |
        |    |      \--- MBC1 Mode Select
        |    \---------- ROM Bank Mask
        \--------------- Multiplex Enable
```

#### Bit 0: MBC1 Mode Select

This is equivalent to the MBC1 Mode Select register.

This 1-bit register selects between the two MBC1 banking modes.

```
00 = Simple ROM Banking Mode (default)

> 01 = RAM Banking Mode / Advanced ROM Banking Mode

Author: rewritten this section in a much better way

In mode 1, the RAM Bank Low register behaves normally. In multiplex mode, this has the side effect of it affecting accesses to the 0000-3FFF region, as per MBC1.

#### Bits 1-5: ROM Bank Mask

_Can only be written in unmapped mode._

If any or all bits are set high, the corresponding bit(s) in the MBC1 ROM banking register are locked in as "game select" bits instead of "banking" bits. Any attempt to write to those bits while its matching bit in this register is 1 is prevented. The lowest bit of the ROM Bank Mask is ignored and treated as always 0, so the lowest bit of ROM Bank Low is always writeable.

Note: changing the mask can alter what bank would be mapped - only the _unmasked_ bits are used for the "bank 0 is treated as bank 1" logic. So if ROM Bank Low is set to 10h, 

#### Multiplex Enable

_Can only be written in unmapped mode._

When set to 1, swaps the functionality of RAM Bank Low with ROM Bank Mid. As RAM Bank Low is writeable in mapped mode, this allows for the contained game to control (up to - if unmasked in "RAM Bank Mask") two extra bits of the full ROM bank number, allowing for larger game ROMs.

> This is equivalent to the "large ROM" wiring of an MBC1 cart.

Author: There's no dedicated section for it in the MBC1 docs - it's noted throughout. I could link the MBC1 doc maybe?

> This is equivalent to the "large ROM" wiring of an MBC1 cart.

Reviewer: In the meantime, I guess. We should have it in the MBC1 docs, though—perhaps it is in #312 ? I haven't had time to look at it yet.

> This is equivalent to the "large ROM" wiring of an MBC1 cart.

Author: It's not in 312. But that can be added :)

## Operation Notes

Because the MMM01 has registers which disable write access to other registers, it is optimal to set the value of certain registers before others. The optimal order is:

```
1. 2000 ROM Bank Register - contains ROM Bank

> 2. 6000 Mode Register     - contains ROM Bank Mask, and Mode

Reviewer: ```suggestion
2. **Mode register** ($6000) — contains ROM Bank Mask, and Mode
```

> 3. 4000 RAM Register      - contains Mode Write Disable, and RAM Bank

Reviewer: ```suggestion
3. **RAM register** ($4000) — contains Mode Write Disable, and RAM Bank
```

> 4. 0000 RAM Enable        - contains RAM Bank Mask, and Mapping Enable

Reviewer: ```suggestion
4. **RAM Enable register** ($0000) - contains RAM Bank Mask, and Mapping Enable
```

# MMM01

> The MMM01 is a mapper specific to multi-game compilation cartridges. It emulates an MBC1 for the contained games, and supports containing a mix of games from KiB ROMs with no RAM, up to the same maximum memory _per game_ as the MBC1:

Author: There's supposed to be a number there

#### Addressing diagrams

When "unmapped":

```
                                    /------------- In a smaller cartridge, only the needed
                                    |              bits are used (e.g 512 KiB uses 19 bits)
                  /---------------------------\
Bits: 22 21 20 19 18 17 16 15 14 13 12 .. 01 00
      \_____________________/  | \____________/
                 |             |      \----------- From Game Boy address
                 |             \------------------ Always 0
                 \-------------------------------- Always 1
```

Mapped, multiplex disabled:

```
Bits: 22 21 20 19 18 17 16 15 14 13 12 .. 01 00
      \___/ \___/ \____________/ \____________/
        |     |          |            \----------- From Game Boy address
        |     |          \------------------------ bitwise and of ROM Bank Low & ROM Bank Mask
        |     \----------------------------------- ROM Bank Mid
        \----------------------------------------- ROM Bank High
```

### 4000-7FFF - ROM Bank $01-7F (Read Only)

On startup (in "unmapped" mode), this is mapped to the second half of the menu program in the last 32 KiB of the ROM.

> When a game is mapped, this area may contain any of the further 16 KiB banks of the game ROM, except game banks $00, $20, $40, or $60.

Reviewer 2: ```suggestion
When a game is mapped, this area may contain any of the further 16 KiB banks of the game ROM, except for game banks $00, $20, $40, or $60.
```

If one of those banks is selected, the low bit is forced to 1 and that bank is mapped instead ($01, $21, $41, or $61).

> i.e. in mapped mode, if `(ROM Bank Low) & ~(ROM Bank Mask)` is equal to $00 (indicating bank $00 within the game ROM), `(ROM Bank Low) | 1` is used instead.

Reviewer 2: We don't use `~` as pseudocode in text. Please review: https://github.com/gbdev/pandocs/wiki/Document-Style

Author: There are in fact no bitwise operations in the pseudocode section at all, if I'm not blind.

As an example, if ROM Bank Low is set to $10, and the ROM Bank Mask is set to $30, then the bank within the game ROM would be `($10) & ~($30) = $00`.
As game bank $00 is disallowed, the low bit is forced on, mapping bank $11 instead of $10.

If [multiplex is enabled[(<#Multiplex Enable>)], the MMM01 has the same limitation as MBC1 regarding accessing game ROM banks $20, $40, and $60 - they can only be mapped to 0000-3FFF (in mode 1), and not to 4000-7FFF.

#### Addressing diagrams

When "unmapped":

```
Bits: 22 21 20 19 18 17 16 15 14 13 12 .. 01 00
      \_____________________/  | \____________/
                 |             |      \----------- From Game Boy address
                 |             \------------------ Always 1? See note
                 \-------------------------------- Always 1
```

> Note: It's suspected the lowest bit of ROM Bank Low (post $00->$01 remapping) still affects the bank mapped to the 4000-7FFF region in "unmapped" mode the same as it does in mapped mode.

Reviewer 2: "suspects" but go in info box reporting "speculation"

      \___/ \___/ \_________/
        |     |        \-------- From Game Boy address
        |     \----------------- bitwise and of RAM Bank Low & RAM Bank Mask
        \----------------------- RAM Bank High
```

In mode 1:

```
Bits: 16 15 14 13 12 .. 01 00
      \___/ \___/ \_________/
        |     |        \-------- From Game Boy address
        |     \----------------- RAM Bank Low (complete value, not masked)
        \----------------------- RAM Bank High
```

## Registers

The MMM01 registers are extensions of the MBC1 registers.
In "mapped" mode, the extra bits cannot be written to, and the MMM01 emulates an MBC1.
In "unmapped" mode, the extra bits are writeable, allowing the menu program to select which game to play and configure its size and RAM access.

All the MMM01 registers are 7-bits in size and set to $00 on power-up.
For the ROM Bank Number register, this behaves as if it was set to $01.

### 0000-1FFF - RAM Enable (Write Only)

```
Bits: X 6 5 4 3 2 1 0
        | \_/ \_____/
        |  |      \----- Bits 0-3: RAM Enable
        |  \------------ Bits 4-5: RAM Bank Mask
        \--------------- Bit 6: Mapping Enable
```

#### Bits 0-3: RAM Enable

> As per MBC1, writing Ah to the lower 4 bits enables the external RAM, and any other value disables it.

Reviewer 2: Hex values must be preceded (prefixed). Please see https://github.com/gbdev/pandocs/wiki/Document-Style
